dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java

```diff
@@ -0,0 +1,167 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.integers.*;
+import com.dat3m.dartagnan.expression.misc.ITEExpr;
+import com.dat3m.dartagnan.expression.processing.ExprTransformer;
+import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.program.event.EventVisitor;
+import com.dat3m.dartagnan.program.event.RegReader;
+import com.dat3m.dartagnan.program.event.RegWriter;
+import com.dat3m.dartagnan.program.event.core.Alloc;
+import com.dat3m.dartagnan.program.event.core.Local;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.dat3m.dartagnan.expression.integers.IntBinaryOp.*;
+import static com.dat3m.dartagnan.expression.integers.IntCmpOp.*;
+
+/*
+    This pass tries to remove unnecessary NULL checks in a very simple manner:
+    it tries to figure out the signedness of expressions and if they are always positive, then they cannot be NULL.
+
+    The signedness is determined as follows:
+      - Registers written to by Allocs as well as MemoryObject expressions are always positive (allocations always succeed!)
+      - Positive constants are positive (duh).
+      - Some operations (like addition) preserve positiveness.
+
+    TODO: The pass is very naive: it globally assigns a signedness to registers rather than per program point.
+ */
+public class RemoveDeadNullChecks implements FunctionProcessor {
+
+    private RemoveDeadNullChecks() { }
+
+    public static RemoveDeadNullChecks newInstance() { return new RemoveDeadNullChecks(); }
+
+    private enum Sign {
+        UNKNOWN,
+        NON_NEG,
+        POS;
+
+        static Sign meet(Sign a, Sign b) {
+            return a.compareTo(b) >= 0 ? b : a;
+        }
+    }
```

Comment:
This means the order of the enum is important, right? If so, we should document this